teps/0060-remote-resource-resolution.md

```diff
@@ -242,207 +251,533 @@ users can leverage those shared pipelines.
-`PipelineRun`. By splitting the resolution logic in to its own external
-machinery we are introducing a barrier between resolution and metadata
-duplication: the "actuation reconcilers" never see the fully-resolved resource,
-only the `spec` passed via the `status`, so they're unable to perform the
-metadata copy. The end result is that the "resolver reconcilers" are also
-responsible for performing this metadata copy.
-
-A lot of the conversation with this TEP so far has revolved around the fact
-that resolution might be something we want to expose not just for Pipelines but
-also other Tekton projects. Having the metadata copy performed by an external
-controller doesn't make a ton of sense in that case: metadata copying is a
-feature/requirement of Tekton Pipelines, not a general requirement for
-resolution.
-
-There are two approaches that I've considered for solutions to this:
-
-1. Change the contract such that the _entire_ resolved YAML document is passed
-   via the `TaskRun` or `PipelineRun` `status`. This would give the Tekton
-   Pipelines reconcilers total flexibility to process the resolved document as
-   they need, but would be backwards-incompatible.
-
-2. Add a `status.resolvedMeta` field to `TaskRuns` and `PipelineRuns` that
-   allows the experimental controller to pass the `metav1.ObjectMeta`
-   information from a `Task` or `Pipeline` alongside the resolved spec.  This
-   would give enough data for Tekton Pipelines' reconcilers to keep metadata
-   copying their responsibility, without being a hard change to the way resolved
-   `specs` are stored on runs.
-
-Given the above two options we are proposing that we employ strategy (1) and
-add a `taskYaml` field to the `status` of `TaskRuns` and a `pipelineYaml` field
-to the status of `PipelineRuns`.
-
-There are some exceptions to the metadata fields that should be copied across.
-For example the annotation `kubectl.kubernetes.io/last-applied-configuration`
-can be extremely large and would bloat a TaskRun or PipelineRun
-during the resolution phase. Similarly for the `metadata.managedFields` field
-The precise set of metadata we include/exclude should be figured out as part
-of this experimental period but a couple of approaches would be:
-
-- Limit to just metadata fields prefixed with `[a-zA-Z].tekton.dev/`.
-- Exclude known edge-case fields (`last-applied-configuration` annotation,
-  `metadata.managedFields`)
-- Limit to just labels.
-
-### Notes/Caveats (optional)
-
-<!--
-What are the caveats to the proposal?
-What are some important details that didn't come across above.
-Go in to as much detail as necessary here.
-This might be a good place to talk about core concepts and how they relate.
--->
+### 1. Establish a new syntax in Tekton Pipelines' `taskRef` and `pipelineRef` structs for remote resources
+
+Add new fields to `taskRef` and `pipelineRef` to set the resolver and
+its parameters. Access to these new fields will be locked behind the
+`alpha` feature gate.
```

Comment:
Nice!

Storage of task and pipeline definitions could be considered as a runtime concern - a `TaskRun` or a `PipelineRun` would ask to execute the definition specified by a certain `Ref`.

When it comes to a `Pipeline` though, the definition of the `Task` used by the `Pipeline` can be considered an author time concern, and that is not well defined unless we specify a `taskRef`.  Today the `taskRef` is only resolvable (unless bundles are used) - when a `Pipeline` is installed on a cluster - which caused a lot of debate when defining how to store `Pipelines` in the catalog.

Having tasks stored in git or container registries makes the definition available independently of the running cluster,  and thus `Pipeline` definitions more self-consistent. 
It may reduce `Pipeline` reusability, however I would argue that if someone wants to run a `Pipeline` from the catalog using same `Tasks` from a different source, that would actually be a different different pipeline, as we have no way today to identify a `Task` as itself globally, independent of the storage location.

Reply:
We could still over time add the ability to override the `taskRef` resolver at runtime?

Reply:
@afrittoli yeah I agree - I don't think anything we're proposing here gets in the way of us implementing a feature like `taskRef` rewriting down the road.  Designing and implementing this in a Resolver first may help us figure out the edge cases before migrating that logic into Pipelines itself.

Reply:
I've been thinking some more about this, and I think there is an issue with this syntax unless we introduce a concept of bundle and bundle resolution. 

Let's consider a repo that contains contains a `Pipeline` that users some `Tasks`, defined alongside the `Pipeline`. 
If we want to use SHAs (as opposed to tags) to references to resources, it's impossible to release / publish the `Pipeline` alongside the `Tasks` in a consistent version, since the `Pipeline` definition must include the version of the `Tasks` that are being published, and that version (in SHA format) is only known after the publishing.

This is true when the target is a container registry but it could also apply to git. If we wanted to create a git tag, we would have to make a PR to update the pipeline definition.

Possible solutions:
- always release tasks and pipelines independently - however that would make the releasing workflow much more complex
- always define the task spec embedded into the `Pipeline`, however we might not want to do that since we cannot define the same metadata with embedded specs and we might want to share the `Task` with other pipelines

The other option is to have a concept of bundle: if a `Pipeline` from a remote resource includes `Tasks` from the same resource, it may refer to them just by name, and the controller would look them up in the same "bundle" i.e. from the same source / version.

We might even consider decoupling the concept of "bundle" from the "container registry" storage, so that we could have bundles with different backends, such as git, object storage or else

Reply:
Yep, that makes total sense! What's the process from here? Do you want to propose a new TEP to start the ball rolling on standardizing that approach?

It would also be possible for a resolver to "figure out" how to rewrite a Pipeline's YAML based on similar rules to the ones you described. E.g. consider the following `pipelineRef` submitted in a `PipelineRun`:

```yaml
pipelineRef:
  resolver: git
  resource:
    repo: github.com/sbwsg/stuff.git
    path: /foo/bar/pipeline.yaml
    commit: abc123
```

The `git` Resolver clones the repo, checks out the commit, and reads `/foo/bar/pipeline.yaml`. Then it "transforms" the PipelineTask `taskRefs`, rewriting any that match the same repo. So this is the pipeline yaml before the transform:

```yaml
kind: Pipeline
# ... metadata etc ...
spec:
  tasks:
  - name: get-code
    taskRef:
      repo: github.com/sbwsg/stuff.git
      path: /foo/baz/git-clone.yaml
```

And this is the pipeline yaml after the transform:

```yaml
kind: Pipeline
# ... metadata etc ...
spec:
  tasks:
  - name: get-code
    taskRef:
      repo: github.com/sbwsg/stuff.git
      path: /foo/baz/git-clone.yaml
      commit: abc123 # <-- This has been copied from the pipelineRef to the taskRef by the git resolver
```

And finally the resolver returns this transformed YAML.

These kinds of features and rules can be implemented per-Resolver until Tekton Pipelines stabilizes on a common syntax for non-OCI bundles per your suggestion.

Reply:
+1

We don't have to make it a mandatory feature of a resolver, but we should at least define how a resolver that supports "bundles" should behave.